qa/tasks/nvmeof.py

```diff
@@ -0,0 +1,397 @@
+        config['ioengine'] = config.get('ioengine', 'sync')
+        config['bsrange'] = config.get('bsrange', '4k-64k')
+        config['numjobs'] = config.get('numjobs', '1')
+        config['size'] = config.get('size', '1G')
+        config['runtime'] = config.get('runtime', '600') # 10 mins
+        config['rw'] = config.get('rw', 'randrw')
+        config['drive'] = self._get_device_path()
+        return config
+
+    def write_fio_config(self, config):
+        """
+        Default:
+        fio --name=nvmeof-fio-test --ioengine=sync --filename /dev/nvme1n1 --rw=randrw /
+            --bsrange=4k-64k --size=1G --numjobs=1 --time_based --runtime=600 --verify=md5 --verify_fatal=1
+        """
+        fio_config=NamedTemporaryFile(mode='w', prefix='fio_nvmeof_', dir='/tmp/', delete=False)
+        fio_config.write('[nvmeof-fio-test]\n')
+        fio_config.write('ioengine={ioe}\n'.format(ioe=config['ioengine']))
+        fio_config.write('bsrange={bsrange}\n'.format(bsrange=config['bsrange']))
+        fio_config.write('numjobs={numjobs}\n'.format(numjobs=config['numjobs']))
+        fio_config.write('size={size}\n'.format(size=config['size']))
+        fio_config.write('time_based=1\n')
+        fio_config.write('runtime={runtime}\n'.format(runtime=config['runtime']))
+        fio_config.write('rw={rw}\n'.format(rw=config['rw']))
+        fio_config.write('filename={nvme_device}\n'.format(nvme_device=config['drive']))
+        fio_config.write('verify=md5\n')
+        fio_config.write('verify_fatal=1\n')
+        fio_config.close()
+        self.remote.put_file(fio_config.name,fio_config.name)
+        return fio_config
+
+    def run_fio(self, config):
+        test_dir = misc.get_testdir(self.ctx) + "/nvmeof_fio_test"
+        fio_version = self.config.get('version', FIO_VERSION_DEFAULT)
+        config = self._set_config_defaults(config)
+        host_name=self.remote.shortname
+        ioengine_pkg = None
+        try:
+            log.info("[nvmeof]: Installing dependencies for fio test on {host_name}".format(host_name=host_name))
+            if config['ioengine'] == 'libaio':
+                system_type = misc.get_system_type(self.remote)
+                ioengine_pkg = 'libaio-devel' if system_type == 'rpm' else 'libaio-dev'
+                log.info("[nvmeof]: Installing {ioengine_pkg} on {system_type}".format(ioengine_pkg=ioengine_pkg, system_type=system_type))
+                install_package(ioengine_pkg, self.remote)
+            fio_config = self.write_fio_config(config)
+
+            fio = "https://github.com/axboe/fio/archive/fio-" + fio_version + ".tar.gz"
+            self.remote.run(args=['mkdir', run.Raw(test_dir),])
+            self.remote.run(args=['cd' , run.Raw(test_dir),
+                             run.Raw(';'), 'wget', fio, run.Raw(';'), run.Raw('tar -xvf fio*tar.gz'), run.Raw(';'),
+                             run.Raw('cd fio-fio*'), run.Raw(';'), './configure', run.Raw(';'), 'make'])
```

Comment:
> I have put the fio version on top as a constant all-caps variable and added an option to specify fio version in job yaml config.

Why is `fio` being built from scratch at all?  Packages should be universally available these days -- it doesn't look like the most recent version is needed here.

It looks like this was copied from `qa/tasks/rbd_fio.py` -- I think the reason it's done there is that in 2015 the RBD engine wasn't packaged ;)  For NVMeOF, `--ioengine=rbd` won't ever be used, so it's moot point.

We should probably get rid of that and just install a package in `qa/tasks/rbd_fio.py` too.

Reply:
I saw `fio` being built from scratch in `qa/tasks/cbt.py` as well (this is also an old commit so it must be for the same reason): 
https://github.com/ceph/ceph/blob/19aa8fc73d267a722c0899bf830bff671b921aa7/qa/tasks/cbt.py#L108

I'll install the packaged fio for the nvmeof task. And maybe open separate PR to change it for `qa/tasks/rbd_fio.py` and `qa/tasks/cbt.py`. Thank you for pointing this out!

Reply:
@idryomov interesting I didn't know we have rbd_fio. Can you say few words about how it is different from fio? And BTW we're looking for running fio on many namespaces in parallel, with data verification, R/W, etc, to test HA. Do you have some examples that we look at how to do it? thx. (@barakda )

Reply:
> Can you say few words about how it is different from fio?

fio has a concept of an I/O engine, which defines the I/O and how it's issued.  Think of it is as interface which gets implemented by multiple backends, which could be file-based, block device-based, network-based or even virtual.  With `--ioengine=sync` requests are issued using regular `read` and `write` syscalls, with `--ioengine=libaio` using `io_submit` and `io_getevents` syscalls and related helpers from libaio, etc.   For `--ioengine=rbd`, it's `rbd_aio_read()` and `rbd_aio_write()` APIs from librbd -- direct to the cluster with no block device or any kind of gateway.

> And BTW we're looking for running fio on many namespaces in parallel, with data verification, R/W, etc, to test HA. Do you have some examples that we look at how to do it?

I think you would need to define several fio jobs.  I'd suggest perusing https://fio.readthedocs.io/en/latest/fio_doc.html for details.